cirq-core/cirq/transformers/stratify.py

```diff
@@ -116,43 +113,101 @@ def _stratify_circuit(
-
-    # Try to move operations to the left to reduce circuit depth, preserving stratification.
-    for curr_idx, moment in enumerate(stratified_circuit):
-        curr_category = curr_idx % num_categories
-        if curr_category == 0:
-            # Moment containing tagged operations to be ignored.
-            continue
-        batch_removals: List[Tuple[int, 'cirq.Operation']] = []
-        batch_inserts: List[Tuple[int, 'cirq.Operation']] = []
+    num_classes = len(classifiers) + 1  # include one "extra" category for ignored operations
+    new_moments: List[List['cirq.Operation']] = []
+
+    # Keep track of the the latest time index addressing a qubit.
+    qubit_time_index = {qubit: -1 for qubit in circuit.all_qubits()}
+
+    # Keep track of the latest time index addressing a measurement/control key.
+    measurement_time_index: Dict['cirq.MeasurementKey', int] = {}
+    control_time_index: Dict['cirq.MeasurementKey', int] = {}
+
+    # The minimum time index for operations with a tag in context.tags_to_ignore.
+    min_time_index_for_ignored_op = 0
+
+    for moment in circuit:
+        # Identify the new time indices that operations should be moved into.
+        ignored_ops = []
+        op_time_indices = {}
         for op in moment:
-            prv_idx = stratified_circuit.earliest_available_moment(op, end_moment_index=curr_idx)
-            prv_category = prv_idx % num_categories
-            should_move_to_next_batch = curr_category < prv_category
-            prv_idx += curr_category - prv_category + num_categories * should_move_to_next_batch
-            assert prv_idx <= curr_idx and prv_idx % num_categories == curr_idx % num_categories
-            if prv_idx < curr_idx:
-                batch_inserts.append((prv_idx, op))
-                batch_removals.append((curr_idx, op))
-        stratified_circuit.batch_remove(batch_removals)
-        stratified_circuit.batch_insert_into(batch_inserts)
-    return drop_empty_moments.drop_empty_moments(stratified_circuit)
+            ignored_op = any(tag in op.tags for tag in context.tags_to_ignore)
+
+            # Get the earliest time index that can accomodate this op, based on its qubits.
+            min_time_index_for_op = max(qubit_time_index[qubit] + 1 for qubit in op.qubits)
+            if ignored_op:
+                min_time_index_for_op = max(min_time_index_for_op, min_time_index_for_ignored_op)
+
+            # Check for any blocking control or measurement operations.
+            for key in protocols.control_keys(op) & measurement_time_index.keys():
+                time_index = measurement_time_index[key]
+                min_time_index_for_op = max(min_time_index_for_op, time_index + 1)
+            for key in protocols.measurement_key_objs(op) & control_time_index.keys():
+                time_index = control_time_index[key]
+                min_time_index_for_op = max(min_time_index_for_op, time_index + 1)
```

Comment:
I factored out the logic for finding the earliest available/accommodating moment, and tentatively placed it in `circuits.circuit.get_earliest_accommodating_moment_index`.  Suggestions welcome for where this function should be moved.

There are currently two related points that I am dissatisfied with:
1. `circuits/circuit.py` currently has to construct `mop_qubits/mkeys/ckeys` twice: (a) in [`_load_contents_with_earliest_strategy`](https://github.com/perlinm/Cirq/blob/090beb63e805a9b7e7a23e3014928940f9161853/cirq-core/cirq/circuits/circuit.py#L1782-L1784), and (b) in [`get_earliest_accomodating_moment_index`](https://github.com/perlinm/Cirq/blob/090beb63e805a9b7e7a23e3014928940f9161853/cirq-core/cirq/circuits/circuit.py#L2760-L2762).
2. Despite using the same method/logic for finding the earliest accommodating moment, `circuits/circuit.py` and `transformers/stratify.py` have to independently update the `qubit/mkey/ckey_index` dictionaries.

Fundamentally, both of these issues stem from the fact that `_load_contents_with_earliest_strategy` has special logic for what to do with `Moment`s rather than `Operation`s, which gets tangled with the logic of identifying an operation's `qubits/mkeys/ckeys`, finding the earliest accommodating moment, and updating the `qubit/mkey/ckey --> index` dictionaries.

Ideas for how to fix these issues?

Reply:
Two possible resolutions (though there may be better ones):
- Make `get_earliest_accommodating_moment_index` accept a `Union[cirq.Operation, cirq.Moment]`, and for moments return something like `max(*qubit_indices.values(), *mkey_indices.values(), *ckey_indices.values())`.  This would make the code relatively "clean", but it would require  to compute a maximum every time (whereas the current implementation keeps track of this maximum "manually").
- Again accept a `cirq.Moment` as an input, but also accept an optional `circuit_length: Optional[int]` argument, and have `get_earliest_accommodating_moment_index` return that `if isinstance(mop, cirq.Moment) and circuit_length is not None`.  This avoids recalculating the circuit length by maximizing over indices, but it is potentially "dangerous", since he method would happily return an invalid circuit length.  On the other hand, garbage in garbage out so maybe that's okay?